### PR TITLE
Feature/jc 555 anime details

### PR DESCRIPTION
**Pull request by EgorRyazanov**

### CODE REVIEW

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<div class="anime">
+	<ng-container *ngIf="anime$ | async as anime">
+		<h2 class="anime__title">{{ anime.titleEnglish | empty }}</h2>
+		<h2 class="anime__title">{{ anime.titleJapanese | empty }}</h2>
+		<div class="anime__details">
+			<img class="details__image" [alt]="anime.titleEnglish" [src]="anime.image" (click)="openDialog(anime.image)" />
```
Comment (AndreySkripachev) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:6]: If there are more than 3 attributes, they should be placed in different lines
```suggestion
			<img 
						class="details__image" 
						[alt]="anime.titleEnglish" 
						[src]="anime.image" 
		        			(click)="openDialog(anime.image)" 
			/>
```

File: apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<mat-dialog-content>
+	<img alt="anime-full-size-image" class="image" [src]="data.imageUrl" />
```
Comment (AndreySkripachev) [inline apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.html:2]: I think the text in the alt attribute will not give the user any information in case the picture fails to load

File: apps/angular/src/app/features/base-page/base-page.component.html [hunk 3]
```diff
@@ -1,5 +1,7 @@
 <header class="header">
-	<h2 class="header__title">AnimeHub</h2>
+	<a class="header__title-link" routerLink="/">
```
Comment (AndreySkripachev) [inline apps/angular/src/app/features/base-page/base-page.component.html:2]: This looks odd to me, since you're putting a block element in a line element. You could put the link inside the header

You can see [here](https://caninclude.glitch.me/caninclude?child=h2&parent=a) why you shouldn't put `h2` tags in `a`

File: apps/angular/src/core/interceptors/auth.interceptor.ts [hunk 4]
```diff
@@ -46,6 +46,7 @@ export class AuthInterceptor implements HttpInterceptor {
 	 * @param userSecret User secret.
 	 */
 	private appendAuthorizationHeader(headers: HttpHeaders, userSecret: UserSecret): HttpHeaders {
+		// console.log9)
```
Comment (AndreySkripachev) [inline apps/angular/src/core/interceptors/auth.interceptor.ts:49]: That doesn't seem like it should be here.

File: apps/angular/src/shared/directives/readable-genres.pipe.ts
Comment (AndreySkripachev) [inline apps/angular/src/shared/directives/readable-genres.pipe.ts:1]: Why did you put the pips in the directives folder?

File: apps/angular/src/shared/directives/readable-source.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+@Pipe({
+	name: 'readableSource',
+})
+export class ReadableSource implements PipeTransform {
+	/**
+	 * Makes source readable.
+	 * @param source Source.
+	 */
+	public transform(source: Sources): string {
+		return Source.toReadable(source);
```
Comment (AndreySkripachev) [inline apps/angular/src/shared/directives/readable-source.pipe.ts:14]: These pipes are very similar to each other and it looks like code duplication. You can make one pipe that will, for example, accept a value and a mapper with the `toReadable(value)` method.

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.css [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  align-items: center;
+  justify-content: center;
+  background-color: var(--background-alternative-primary-color);
+}
+
+.anime__details {
+  display: flex;
+}
+
+.details__image {
```
Comment (maxineo) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.css:20]: I see no `details` block. Perhaps it should be `anime__details-image`?

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.css [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  max-width: 350px;
+  max-height: 350px;
+  float: left;
+  margin-right: var(--space-md);
+  margin-bottom: var(--space-md);
```
Comment (maxineo) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.css:25]: Try to position elements from parent. That way it will be a lot easier to reuse them. [wiki](https://wiki.saritasa.rocks/frontend/languages/css#componentselements-should-not-position-themselves)

File: apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
+import { MatButtonModule } from '@angular/material/button';
+import { MAT_DIALOG_DATA, MatDialogModule } from '@angular/material/dialog';
+
+/** Image dialog data. */
+interface ImageDialogData {
+
+	/** Image URL. */
+	imageUrl: string;
```
Comment (maxineo) [inline apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.ts:9]: Missing `readonly`.

File: apps/angular/src/shared/directives/readable-source.pipe.ts (hunk 5 above)
Comment (maxineo) [reply apps/angular/src/shared/directives/readable-source.pipe.ts:14]: I think we can just use generic type.

File: libs/core/dtos/anime-dto/anime-details.dto.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+	/** Studios. */
+	readonly studios_data: readonly StudioDTO[];
+
+	/** Genres. */
+	readonly genres_data: readonly GenreDto[];
+}
+
+/** Seasons DTO. */
+export enum SeasonsDTO {
```
Comment (maxineo) [inline libs/core/dtos/anime-dto/anime-details.dto.ts:67]: Why dto is written with all uppercase letters here and in the next enum while it has only first uppercasse letter in other cases?

File: libs/core/dtos/anime-dto/anime-details.dto.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+	Summer = 'SUMMER',
+	Winter = 'WINTER',
+	Spring = 'SPRING',
+	Fall = 'FALL',
+	NonSeasonal = 'NON_SEASONAL',
+}
+
+export enum SourceDTO {
```
Comment (maxineo) [inline libs/core/dtos/anime-dto/anime-details.dto.ts:75]: Missing comment.

File: libs/core/dtos/anime-dto/anime-details.dto.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
+import { StudioDTO } from '../studios-dto/studio.dto';
+import { GenreDto } from '../genre-dto/genre.dto';
+
+import { AnimeDtoTypes, AnimeStatusDto } from './anime.dto';
+
+export interface AnimeDetailDto {
```
Comment (maxineo) [inline libs/core/dtos/anime-dto/anime-details.dto.ts:6]: Missing comment.

File: libs/core/dtos/studios-dto/studio.dto.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+export interface StudioDTO {
```
Comment (maxineo) [inline libs/core/dtos/studios-dto/studio.dto.ts:1]: Missing documentation.

File: libs/core/models/anime/anime-detail.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+	 *  Start date.
+	 *  @example 1975-01-01T00:00:00Z.
+	 */
+	start: Date | null;
+
+	/**
+	 *  End date.
+	 *  @example 1975-01-01T00:00:00Z.
+	 */
+	end: Date | null;
```
Comment (maxineo) [inline libs/core/models/anime/anime-detail.ts:106]: Is there a reason why these are not `readonly`?

**Review by maxineo — COMMENTED** [review]
There are several undocumented exported members across all files, so please revise them.

File: apps/angular/src/shared/pipes/yes-no.pipe.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+/** Yes or no pipe. */
+@Pipe({
+	name: 'yesOrNo',
+})
+export class YesNo implements PipeTransform {
```
Comment (dmitrymaslov13) [inline apps/angular/src/shared/pipes/yes-no.pipe.ts:7]: Angular entities usually have an ending that describes what they mean.
In this case expected `YesNoPipe` class name

File: apps/angular/src/shared/pipes/readable.pipe.ts [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+import { Season, Seasons } from '@js-camp/core/models/season';
+import { AnimeStatus, AnimeStatuses } from '@js-camp/core/models/anime/anime-status';
+import { GenreMapper } from '@js-camp/core/mappers/genre/genre.mapper';
+import { isType } from '@js-camp/core/utils/is-type';
+
+/** Readable pipe. */
+@Pipe({
+	name: 'readable',
+})
+export class Readable implements PipeTransform {
```
Comment (dmitrymaslov13) [inline apps/angular/src/shared/pipes/readable.pipe.ts:17]: Same about naming, but please save your time.
I understand what your decision is dictated by, but this approach has several problems

1) Lack of typing for incoming parameters.
2) Non-obvious behavior when passing a type that is not taken into account, it is not very obvious that the same thing will be returned as passed before
3) The need to enter into this file all new types that will appear in the future (we don't want to be non-consistent :D)
4) Very unreadable transform method with infinite `if` `else`

Have you heard of the **god object** antipattern?
In short, it's about an object that stores a lot of things, or does a lot of different things

It violates **S** from **SOLID** (S - Single Responsibility)
which in short says that each method / class should be responsible for its own responsibility

In your case class have responsible for map to readable Source/Rating/AnimeType/... (more than one, right? :D )

Let's transform this into small and beautiful pipes, each of which is responsible for its own type.

And i advise you to read about **SOLID** to keep in mind in the process what it means

**Review by dmitrymaslov13 — CHANGES_REQUESTED** [review]
And your demo not working)
I not can open anime details

File: apps/angular/src/shared/pipes/readable.pipe.ts (hunk 15 above)
Comment (dmitrymaslov13) [reply apps/angular/src/shared/pipes/readable.pipe.ts:17]: I saw other guys comments about this.
@maxineo, @AndreySkripachev what do you think about it? Maybe did you mean something else?

File: apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+@Component({
+	selector: 'camp-image-dialog',
+	templateUrl: './image-dialog.component.html',
+	styleUrls: ['./image-dialog.component.css'],
+	standalone: true,
+	changeDetection: ChangeDetectionStrategy.OnPush,
+	imports: [MatDialogModule, MatButtonModule],
+})
+export class ImageDialogComponent {
+	public constructor(@Inject(MAT_DIALOG_DATA) public data: ImageDialogData) {}
```
[the remark this answers is not in the document]
Comment (dmitrymaslov13) [reply apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.ts:25]: Just for you know: 

i see you use `inject` sometimes.
You can use `inject` for inject `MAT_DIALOG_TOKEN` token.

```suggestion
	protected readonly data = inject<ImageDialogData>(MAT_DIALOG_DATA);
```

File: apps/angular/src/app/features/anime/animes-page/animes-page.component.css [hunk 17]
```diff
@@ -37,5 +37,13 @@
...
   z-index: 100;
-  background-color: var(--spinner-background-color);
+  background-color: var(--background-alternative-primary-color);
+}
+
+.table__row {
+  cursor: pointer;
+}
+
+tr.table__row:hover {
```
[the remark this answers is not in the document]
Comment (EgorRyazanov) [reply apps/angular/src/app/features/anime/animes-page/animes-page.component.css:47]: It's does not working like this, idk why, i tried few variants, but stackoverflow gave me such answer

File: apps/angular/src/core/services/anime.service.ts [hunk 18]
```diff
@@ -8,15 +8,17 @@ import { AnimeParametersMapper } from '@js-camp/core/mappers/anime/anime-params.
...
+import { AnimeDetailMapper } from '@js-camp/core/mappers/anime/anime-details.mapper';
+import { AnimeDetail } from '@js-camp/core/models/anime/anime-detail';
+
 import { UrlService } from './url.service';
 
 /** Anime service. */
-@Injectable({
-	providedIn: 'root',
-})
+@Injectable()
```
[the remark this answers is not in the document]
Comment (EgorRyazanov) [reply apps/angular/src/core/services/anime.service.ts:18]: I figured that this is the anime service that will perform logic exclusively in anime module. Why not?

File: apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.ts [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+/** Image dialog. */
+@Component({
+	selector: 'camp-image-dialog',
+	templateUrl: './image-dialog.component.html',
+	styleUrls: ['./image-dialog.component.css'],
+	standalone: true,
+	changeDetection: ChangeDetectionStrategy.OnPush,
+	imports: [MatDialogModule, MatButtonModule],
```
[the remark this answers is not in the document]
Comment (EgorRyazanov) [reply apps/angular/src/app/features/anime/anime-details-page/components/dialog/image-dialog.component.ts]: I forgot to fix it when I was doing it

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+				allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
+				title="YouTube video player"
+				[src]="url"
+			></iframe>
+		</ng-container>
+	</ng-container>
+</div>
+<div *ngIf="isLoading$ | async" class="anime__spinner">
+	<camp-spinner />
+</div>
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:41]: Can we avoid redundant `div` and set structural directive with class directly to `camp-spinner`?
```suggestion
<camp-spinner *ngIf="isLoading$ | async" class="anime__spinner" />
```

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 21]
```diff
@@ -0,0 +1,41 @@
+<div class="anime">
+	<ng-container *ngIf="anime$ | async as anime">
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:2]:
```suggestion
<div class="anime" *ngIf="anime$ | async as anime">
```

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+		<h2 class="anime__title">{{ anime.titleEnglish | empty }}</h2>
+		<h2 class="anime__title">{{ anime.titleJapanese | empty }}</h2>
+		<div class="anime__details">
+			<img
+				class="anime__details-image"
+				[alt]="anime.titleEnglish"
+				[src]="anime.imageUrl"
+				(click)="openDialog(anime.imageUrl, anime.titleEnglish)"
+			/>
+			<div>
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:12]: Please, revise your HTML, maybe we could make it better and write more semantic HTML? For example, it can be `article` instead `div`. Don't forget about UX and semantic when your app has content.

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+				<p><strong>Season:</strong> {{ anime.season | readable }}</p>
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:13]: The `<strong>` HTML element indicates that its contents have strong importance, seriousness, or urgency. So, if you added it only to **bold** text, let's use CSS.

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 24]
```diff
@@ -0,0 +1,41 @@
+<div class="anime">
+	<ng-container *ngIf="anime$ | async as anime">
+		<h2 class="anime__title">{{ anime.titleEnglish | empty }}</h2>
+		<h2 class="anime__title">{{ anime.titleJapanese | empty }}</h2>
+		<div class="anime__details">
+			<img
+				class="anime__details-image"
+				[alt]="anime.titleEnglish"
+				[src]="anime.imageUrl"
+				(click)="openDialog(anime.imageUrl, anime.titleEnglish)"
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:10]: If you do some element interactive, don't forget to check - does this element have access from keyboard?

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+				<p><strong>Date start:</strong> {{ anime.aired.start | date | empty }}</p>
+				<p><strong>Date end:</strong> {{ anime.aired.end | date | empty }}</p>
+				<p><strong>Genres:</strong> {{ anime.genresData | readable | empty }}</p>
+				<p><strong>Studios:</strong> {{ anime.studiosData | readable | empty }}</p>
+				<p><strong>Rating:</strong> {{ anime.rating | readable | empty }}</p>
+				<p><strong>Source:</strong> {{ anime.source | readable | empty }}</p>
+				<p><strong>Type:</strong> {{ anime.type | readable | empty }}</p>
+				<p><strong>Status:</strong> {{ anime.status | readable | empty }}</p>
+				<p><strong>Created:</strong> {{ anime.created | date | empty }}</p>
+				<p><strong>Modified:</strong> {{ anime.modified | date | empty }}</p>
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:25]: Revise formatting, please.
```suggestion
				<p><strong>Modified:</strong>{{ anime.modified | date | empty }}</p>
```

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+			</div>
+		</div>
+		<ng-container *ngIf="saveVideoUrl$ | async as url">
+			<iframe
+				allowfullscreen
+				class="anime__video"
+				allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
+				title="YouTube video player"
+				[src]="url"
+			></iframe>
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html:35]: What if we make a standalone component which purpose to show such content? What do you think?

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+/** Anime details page. */
+@Component({
+	selector: 'camp-anime-details-page',
+	templateUrl: './anime-details-page.component.html',
+	styleUrls: ['./anime-details-page.component.css'],
+	changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class AnimeDetailsPageComponent {
+	/** ID. */
+	protected readonly id$: Observable<string>;
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts:23]: Why does this property have `protected` modifier? I don't see where you are using it on template.

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+	private createIdParamStream(): Observable<string> {
+		return this.activeRoute.paramMap.pipe(map(params => params.get('id') ?? ''));
+	}
+
+	/** Creates anime stream. */
+	private createAnimeStream(): Observable<AnimeDetail> {
+		return this.id$.pipe(
+			tap(() => {
+				this.isLoading$.next(true);
+			}),
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts:75]:
```suggestion
			tap(() => this.isLoading$.next(true)),
```

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+			tap(animeDetail => {
+				if (animeDetail.trailerYoutubeUrl != null) {
+					this.saveVideoUrl$.next(this.sanitizer.bypassSecurityTrustResourceUrl(animeDetail.trailerYoutubeUrl));
+				}
+			}),
+			catchError((error: unknown) => {
+				this.router.navigate([homeUrl]);
+				return throwError(() => error);
+			}),
+			stopLoadingStatus(this.isLoading$),
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts:86]: If you have `stopLoadingStatus`, maybe it would be better to create `startLoadingStatus` or something else?

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+			tap(() => {
+				this.isLoading$.next(true);
+			}),
+			switchMap(id => this.animeDetailsService.getAnime(id)),
+			tap(animeDetail => {
+				if (animeDetail.trailerYoutubeUrl != null) {
+					this.saveVideoUrl$.next(this.sanitizer.bypassSecurityTrustResourceUrl(animeDetail.trailerYoutubeUrl));
+				}
+			}),
+			catchError((error: unknown) => {
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts:82]: Maybe, we notify user about what happened? For example, show dialog or something like that.

File: apps/angular/src/app/features/anime/animes-page/animes-page.component.css (hunk 17 above)
[the remark this answers is not in the document]
Comment (viktor-sakhno-saritasa) [reply apps/angular/src/app/features/anime/animes-page/animes-page.component.css:46]: It doesn't work due to selector specificity)

File: apps/angular/src/shared/pipes/readable.pipe.ts [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+	 */
+	private isAnimeStatusesType(type: string): type is AnimeStatuses {
+		return isType(type, AnimeStatuses);
+	}
+
+	/**
+	 * Makes unknown readable.
+	 * @param value Value to transform.
+	 */
+	public transform(value: readonly unknown[] | string): readonly unknown[] | string {
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/shared/pipes/readable.pipe.ts:78]: I think it would be better to create several separates pipes to keep single responsibility approach. At the moment, this class is very fragile, and it will change every time when you want to add a new entity.

File: apps/angular/src/shared/pipes/readable.pipe.ts [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+		if (value instanceof Array) {
+			if (this.isStudioArrayType(value)) {
+				return StudioMapper.toReadable(value);
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/shared/pipes/readable.pipe.ts:81]: Also, `toReadable` isn't a task for mappers. It is a UI related feature. So, I think we can add `toReadable` to `Studio` namespace or something like that.

File: libs/core/dtos/anime-dto/anime-details.dto.ts [hunk 33]
```diff
@@ -0,0 +1,108 @@
+import { StudioDto } from '../studios-dto/studio.dto';
+import { GenreDto } from '../genre-dto/genre.dto';
+
```
Comment (viktor-sakhno-saritasa) [inline libs/core/dtos/anime-dto/anime-details.dto.ts:3]: Remove empty line, pls.

File: libs/core/mappers/aired.mapper.ts [hunk 34]
```diff
@@ -0,0 +1,17 @@
+import { AiredDto } from '../dtos/aired-dto';
+import { Aired } from '../models/anime/anime-detail';
+
+/** Aired. */
+export namespace Airing {
```
Comment (viktor-sakhno-saritasa) [inline libs/core/mappers/aired.mapper.ts:5]: Usually, we sync our file name and main class what we export. So, if you have mapper for `aired` entity, let's name our namespace like `AiredMapper`.

**Review by viktor-sakhno-saritasa — CHANGES_REQUESTED** [review]

File: apps/angular/src/shared/pipes/readable.pipe.ts (hunk 31 above)
Comment (pavelzadorin-saritasa) [reply apps/angular/src/shared/pipes/readable.pipe.ts:78]: I completely agree, remember I said about code smells, look: https://refactoringguru.cn/smells/switch-statements
Think, or let's discuss how you can replace it.

**Review by pavelzadorin-saritasa — CHANGES_REQUESTED** [review]

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html (hunk 25 above)
Comment (EgorRyazanov) [reply apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.html]: I need this space.
With space: 
image
Without space:
image

File: apps/angular/src/app/features/anime/animes-page/animes-page.component.ts [hunk 35]
```diff
@@ -48,6 +52,12 @@ export class AnimesPageComponent {
 	/** Status of anime. */
 	protected readonly isLoading$ = new BehaviorSubject(false);
```
Comment (khoanguyen-saritasa) [inline apps/angular/src/app/features/anime/animes-page/animes-page.component.ts:53]: Wrong docs.

**Review by khoanguyen-saritasa — APPROVED** [review]

File: apps/angular/src/app/features/base-page/base-page.component.ts [hunk 36]
```diff
@@ -19,13 +20,20 @@ export class BasePageComponent {
 	/** Destroy ref. */
 	private readonly destroyRef = inject(DestroyRef);
 
+	/** Router. */
```
Comment (viktor-sakhno-saritasa) [inline apps/angular/src/app/features/base-page/base-page.component.ts:23]: It's not required to comment private properties, especially, such properties like `router`, `destroyRef`, etc. It's clear.

**Review by viktor-sakhno-saritasa — APPROVED** [review]

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts [hunk 37]
```diff
@@ -0,0 +1,122 @@
...
+/** Anime details page. */
+@Component({
+	selector: 'camp-anime-details-page',
+	templateUrl: './anime-details-page.component.html',
+	styleUrls: ['./anime-details-page.component.css'],
+	changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class AnimeDetailsPageComponent {
+	/** ID. */
+	private readonly id$: Observable<string>;
```
Comment (dbitkovski) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts:33]: Why do you need separate stream for id? I think you can get it within `createAnimeStream` method

File: apps/angular/src/shared/pipes/yes-no.pipe.ts [hunk 38]
```diff
@@ -0,0 +1,15 @@
...
+@Pipe({
+	name: 'yesOrNo',
+})
+export class YesNoPipe implements PipeTransform {
+	/**
+	 * Converts boolean to yes or no.
+	 * @param value Boolean type value.
+	 */
+	public transform(value: boolean): 'Yes' | 'No' {
+		return value ? 'Yes' : 'No';
```
Comment (dbitkovski) [inline apps/angular/src/shared/pipes/yes-no.pipe.ts:13]: What if value is `null` or `undefined`? I believe ts type-checking doesn't help here

**Review by dbitkovski — APPROVED** [review]

File: apps/angular/src/app/features/anime/animes-page/animes-page.component.html [hunk 39]
```diff
@@ -50,14 +50,19 @@
 			</ng-container>
 			<ng-container matColumnDef="status">
 				<th mat-sort-header="status" mat-header-cell *matHeaderCellDef>Status</th>
-				<td mat-cell *matCellDef="let anime">{{ anime.status | readableAnimeStatus }}</td>
+				<td mat-cell *matCellDef="let anime">{{ animeStatus.toReadable(anime.status) }}</td>
```
[the remark this answers is not in the document]
Comment (EgorRyazanov) [reply apps/angular/src/app/features/anime/animes-page/animes-page.component.html:53]: I was talking about this with @pavelzadorin-saritasa and @dmitrymaslov13. They said its extra wrapper. Maybe I misunderstood them

File: libs/core/models/anime/anime-type.ts [hunk 40]
```diff
@@ -27,7 +27,11 @@ export namespace AnimeTypes {
...
 	 * @param value Anime types.
 	 */
-	export function toReadable(value: AnimeType): string {
-		return TO_TITLE_MAP[value];
+	export function toReadable(value: AnimeType | null): string | null {
+		if (value != null) {
+			return TO_TITLE_MAP[value];
+		}
+
+		return null;
```
Comment (dmitrymaslov13) [inline libs/core/models/anime/anime-type.ts:35]:
```suggestion
	export function toReadable(value: AnimeType): string {
		return TO_TITLE_MAP[value];
```

File: libs/core/models/season.ts [hunk 41]
```diff
@@ -0,0 +1,31 @@
...
+	 * Makes season readable.
+	 * @param season Season.
+	 */
+	export function toReadable(season: Season | null): string | null {
+		if (season != null) {
+			return TO_TITLE_MAP[season];
+		}
+
+		return null;
+	}
```
Comment (dmitrymaslov13) [inline libs/core/models/season.ts:30]:
```suggestion
	export function toReadable(season: Season): string {
		return TO_TITLE_MAP[season];
	}
```

File: libs/core/models/anime/anime-status.ts [hunk 42]
```diff
@@ -19,7 +19,11 @@ export namespace AnimeStatus {
...
 	 * @param value Anime status.
 	 */
-	export function toReadable(value: AnimeStatus): string {
-		return TO_TITLE_MAP[value];
+	export function toReadable(value: AnimeStatus | null): string | null {
+		if (value != null) {
+			return TO_TITLE_MAP[value];
+		}
+
+		return null;
```
Comment (dmitrymaslov13) [inline libs/core/models/anime/anime-status.ts:27]:
```suggestion
	export function toReadable(value: AnimeStatus): string {
		return TO_TITLE_MAP[value];
```

File: libs/core/models/anime/anime-source.ts [hunk 43]
```diff
@@ -0,0 +1,55 @@
...
+	 * Makes source readable.
+	 * @param source Source.
+	 */
+	export function toReadable(source: Source | null): string | null {
+		if (source != null) {
+			return TO_TITLE_MAP[source];
+		}
+
+		return null;
+	}
```
Comment (dmitrymaslov13) [inline libs/core/models/anime/anime-source.ts:54]:
```suggestion
	export function toReadable(source: Source): string {
		return TO_TITLE_MAP[source];
	}
```

File: apps/angular/src/shared/components/video-player/video-player.component.ts [hunk 44]
```diff
@@ -0,0 +1,28 @@
...
+export class VideoPlayerComponent {
+	/**
+	 * Video URL.
+	 * @param url Video URL.
+	 */
+	@Input({ required: true })
+	public set videoUrl(url: string | null) {
+		if (url != null) {
+			this.saveVideoUrl$.next(this.sanitizer.bypassSecurityTrustResourceUrl(url));
+		}
```
Comment (dmitrymaslov13) [inline apps/angular/src/shared/components/video-player/video-player.component.ts:21]: This component looks like a reusable component, but have problem with pass input `null` after pass `string` url, video will not be lost

File: apps/angular/src/core/services/error-dialog.service.ts [hunk 45]
```diff
@@ -0,0 +1,23 @@
+import { Injectable, inject } from '@angular/core';
+import { MatDialog, MatDialogRef } from '@angular/material/dialog';
+import { ErrorDialogComponent } from '@js-camp/angular/shared/components/error-dialog/error-dialog.component';
+
+/** Confirm dialog service. */
+@Injectable({
+	providedIn: 'root',
+})
+export class ErrorDialogService {
+	private readonly errorDialog = inject(MatDialog);
```
Comment (dmitrymaslov13) [inline apps/angular/src/core/services/error-dialog.service.ts:10]: the name is somewhat misleading.
it is a tool to open any material dialogs

File: apps/angular/src/core/services/auth-api.service.ts [hunk 46]
```diff
@@ -1,15 +1,15 @@
 import { HttpClient } from '@angular/common/http';
 import { Injectable, inject } from '@angular/core';
-import { Observable } from 'rxjs';
-import { map } from 'rxjs/operators';
-import { Registration } from '@js-camp/core/models/auth/registration';
-import { UserSecret } from '@js-camp/core/models/auth/user-secret';
-import { UserSecretMapper } from '@js-camp/core/mappers/auth/user-secret.mapper';
-import { Login } from '@js-camp/core/models/auth/login';
+
```
Comment (dmitrymaslov13) [inline apps/angular/src/core/services/auth-api.service.ts:3]:
```suggestion
```

File: apps/angular/src/core/services/anime.service.ts [hunk 47]
```diff
@@ -15,23 +18,31 @@ import { UrlService } from './url.service';
...
 	}
+
+	/**
+	 * Gets anime by ID.
+	 * @param id ID of anime.
+	 */
+	public getAnime(id: string): Observable<AnimeDetail> {
+		return this.httpService
+			.get<AnimeDetailDto>(this.urlService.animeUrls.animesDetail(id))
+			.pipe(map(dto => AnimeDetailMapper.fromDto(dto)));
```
Comment (dmitrymaslov13) [inline apps/angular/src/core/services/anime.service.ts:46]: Just for you know: 

I see that errors in such requests are not processed in the application.

But I would like you to keep in mind that any errors on the domain layer should refer to the domain layer.

If this request throws an error and you try to catch that error in the application now, you will catch a server error, not an domain error.

You can mark a comment as resolved, or embed mappers for domain errors in requests. The purpose of this comment is just to tell you about it

File: apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts [hunk 48]
```diff
@@ -0,0 +1,113 @@
...
+	private createAnimeStream(): Observable<AnimeDetail> {
+		return this.id$.pipe(
+			startLoadingStatus(this.isLoading$),
+			switchMap(id => this.animeDetailsService.getAnime(id)),
+			catchError((error: unknown) => {
+				if (error instanceof HttpErrorResponse) {
+					this.errorDialogService.openDialog(error.message);
+				}
+				this.router.navigate([homeUrl]);
+				return throwError(() => error);
```
Comment (dmitrymaslov13) [inline apps/angular/src/app/features/anime/anime-details-page/anime-details-page.component.ts:108]: Oops, I missed this point in the comments about the errors.
Abstraction emissions have occurred here
We should not work with server errors and catch them at the domain and UI layers!

**Review by dmitrymaslov13 — CHANGES_REQUESTED** [review]

**Review by dmitrymaslov13 — APPROVED** [review]

**Review by pavelzadorin-saritasa — APPROVED** [review]
